middleware/faq.rst

```diff
@@ -535,6 +535,251 @@ The resulting query logs can be useful for developers and database administrator
 
 .. _middleware_faq_clean_up_runs:
 
+How can I get a report on all failures and missing datasets in a run?
+=====================================================================
+.. _middleware_faq_pipetask_report:
+
+The :any:`pipetask report <lsst.ctrl.mpexec-pipetask#report>` tool can be used to analyze executed quantum graphs, troubleshoot, diagnose failures, and confirm that fail-and-recovery attempts (such as when using ``--skip-existing-in``) are effective.
+
+When analyzing multiple graphs with ``pipetask report``, all graphs should be attempts to execute the same pipeline with the same dataquery.
+
+The recommended usage is
+
+.. code:: sh
+
+  pipetask report --full-output-filename <path/to/output_file>.json --force-v2 REPO QGRAPHS
+
+
+- The ``<path/to/output_file>.json`` option is the path to a file where the output can be stored
+- The ``--force-v2`` option makes sure that the most recent version of the tool is used even when the user passes only one graph
+- The ``REPO`` argument is the `Butler` repo where the output from the processing is stored
+- The ``QGRAPHS`` argument is a ``Sequence`` of `QuantumGraph`s to be analyzed, separated by spaces and passed in order of first to last executed
+
+This will print two ``bps report``-style tables, one for quanta and one for output datasets.
+
+In the output JSON file will be
```

Comment:
I tested this locally:
```
SUBMITDIR=/sdf/group/rubin/user/lskelvin/bps/submit/u/lskelvin/DM-45602/step1/20240808T015002Z

pipetask report \
--full-output-filename $SUBMITDIR/u_lskelvin_DM-45602_step1_20240808T015002Z.info.json \
--force-v2 \
$REPOMAIN \
$SUBMITDIR/u_lskelvin_DM-45602_step1_20240808T015002Z.qgraph
```

which did indeed produce only the two tables to `stdout` I was expecting. However, my existing `.json` file had its timestamp updated - is this to be expected?

Reply:
Hmm. Is there anything in the json file? (there should be! the contents of a full `QuantumProvenanceGraph.Summary` Pydantic model, specifically.)

Pipetask report will save its output in a json file if you pass one via `--full-output-filename`. I generally suggest it! It's got slightly more information, is a good way to save the info in the tables, and if you do have multiple submits for multiple data queries that you want to combine over the same "step" (using `pipetask aggregate-reports`) you'll need the file later.

When you save this info to a file, `pipetask report` still outputs tables following the `--brief` format (no error summaries after the tables unless something is wonky or cursed). So you get your counts of quanta on stdout but your long-form "this data id had this failure" in the json file which you can count up and search.

This mostly happens in `report_v2` and `print_summary` here and below: https://github.com/lsst/ctrl_mpexec/blob/3abcffb608b3ac821373b65250b5f20c81ebc6dd/python/lsst/ctrl/mpexec/cli/script/report.py#L118 (It's got tons of comments & helpful docstrings; the documentation should show up [here](https://pipelines.lsst.io/v/weekly/modules/lsst.ctrl.mpexec/pipetask.html#pipetask-report) whenever newly-deployed code updates).

I'll think about phrasing the intro bit so that it's clear that my suggested usage is to output a file. Let me know if you have any other suggestions after reading this comment :)

Reply:
Ah, I get it now! I thought that `--full-output-filename <path/to/output_file>.json` meant the `.json` file output by my BPS run, and _not_ the file that gets output by `pipetask report`! So, for my test, I provided the path to the existing `.json` file provided by BPS, which then just got overwritten by the `pipetask report` `.json` file when I ran the command :facepalm:.

Reply:
Ahh, sorry if you needed something in there! This is useful for me to know, so I can make sure this is all phrased in a helpful way.

Reply:
To help someone like me _not_ do this next time, perhaps this line:

```
- The ``<path/to/output_file>.json`` option is the path to a file where the output can be stored
```

can be changed to:
```
- The ``--full-output-filename <path/to/output_file>.json`` option provides the path to a file where the output can be stored
```

and then, below:

```
.. note::

    If the ``--full-output-filename`` argument is not provided, the contents of a full ``QuantumProvenanceGraph.Summary`` Pydantic model will be output to the command line instead.
```

Reply:
Also, perhaps a future improvement for `pipetask report` (beyond the scope of this ticket), it would be nice if the task and datasets tables printed _after_ the Pydantic model information, to make it so that I don't have to scroll up very far to see the tabular data?

Reply:
Oh, and another note too: if the `<path/to/output_file>.json` file already exists, it would be great if `pipetask report` warned me of that and aborted with an error along the lines of "Pydantic model `.json` file already exists, and `--overwrite` has not been provided", and then adding in a new `--overwrite` argument which forces that overwrite operation. We have something similar in the `source_injection` `make_injection_pipeline` tool, which has helped prevent unexpected file overwrites (as I did in my initial testing of `pipetask report`).

Reply:
I expect many end-users will not even need the output Pydantic model information, and are merely looking for the summary tables. In which case, if `--full-output-filename` isn't provided, the Pydantic model info isn't printed at all. I'm not sure if there's a good reason either for `--filename` or `--output` as opposed to `--full-output-filename`? Anyway, I hope this has been useful feedback!

Reply:
Ooh, I could probably do something like `--output-file` -- `--output` feels a bit too close to an output collection. The "Just tables" option is called `--brief` -- which is also the version you see in the terminal if you save a file.

Reply:
Right! I guess I'm advocating for `--brief` to be the default, and another argument to be `--pydantic`, or a similar, better name. But, you liaise more with others who use this tool, so I'd be keen to get their thoughts too.

Reply:
I agree that we should not be using the term "pydantic" in user facing interfaces.

Reply:
Right. I just don't have a good name to describe the output given by `--full-output-filename`, and needed something to communicate my meaning above.

Reply:
Hm, yeah. I can see why `--brief` only might be a good option for many. I wanted to fight the possible impression that this tool _doesn't_ output the list of error messages with their data ids (really the feature that makes it stand out compared to `bps report`). I'll have a good think on this, though. A follow-up ticket to make the tables print _after_ the list seems good.